Add support for Buffer and GeoPoint data types

hideya opened the pull request:
This fixes Issue https://github.com/strongloop/loopback-sdk-ios/issues/60

`Buffer` is mapped to either of `NSData` or `NSMutableData` (reflecting a specified property definition in a subclass of `LBPersistedModel`).

`GeoPoint` is mapped to `CLLocation`.  Initially I thought to introduce a new class like `LBGeoPoint` (a straight translation of LoopBack `GeoPoint` class) but took a simpler approach as the reference of `-[CLLocation initWithLatitude:longitude:]` says "typically, you acquire location objects from the location service, but you can use this method to create new location objects for other uses in your application."  A downside of this approach is a possible confusion when a user tried to store and retrieve a `CLLocation` object, as `GeoPoint` stores latitude and longitude only whereas `CLLocation` has other properties including altitude, accuracies, timestamp, etc.  However, I thought that the benefit of reusing `CLLocation` (simplicity in usage) would be greater.  BTW, `CLLocationCoordinate2D` struct was also considered for mapping, but complications in its use were found it as it is not a class.

This PR also includes the following changes:
- Fix a race condition found in -[LBModel initWithRepository:...].
- Add typedef LBPersistedModelDictionarySuccessBlock which is needed by CodeGen.
- Fix indentations/format of LBPersistedModelTests.m for readability.

@bajtos would you please review the changes?

bajtos commented:
Hi @hideya, thank you for the pull request. I'm quite overloaded these days, I'll try to review this soon.

bajtos commented:
Hi @hideya, I am afraid your implementation of `Buffer` is not compatible with what we already have in strong-remoting, see [serialization](https://github.com/strongloop/strong-remoting/blob/e0846fe0f9d5480eb3e2ec5105b5fe6b4b7fa332/lib/shared-method.js#L548-L552) and [deserialization](https://github.com/strongloop/strong-remoting/blob/e0846fe0f9d5480eb3e2ec5105b5fe6b4b7fa332/lib/shared-method.js#L490-L500).

```js
cb(null, new Buffer([50, 51])); }
// produces
{ '$type': 'base64', '$data': 'MjM=' }
```

Your current test suite verifies that the iOS client can parse the data it sent before, it does not verify whether server method will receive the data as buffer and whether a buffer produced by server is correctly parsed by the client. Could you please add these two missing tests?

bajtos commented on LoopBack/LBModel.m:
```diff
@@ -22,6 +21,83 @@ - (NSMutableDictionary *)_overflow;
+
++ (CLLocation *)locationFromJSONObject:(NSDictionary *)jsonObject {
+    if ([jsonObject count] != 2 || jsonObject[@"lat"] == nil || jsonObject[@"lng"] == nil) {
+        return nil;
```
I think this will be difficult to troubleshoot. Could you please add a debug log or some kind of a warning, to tell the developer that invalid data was received for a location property?

hideya replied:
Good point.  I will add a debug log there.  I'll also add a log output to `+[NSMutableData (LBModelAdditions) dataFromJSONObject:]` too.

raymondfeng replied:
I am out of the office until 01/04/2016.

Happy holidays!


Note: This is an automated response to your message  "Re:
[loopback-sdk-ios] Add support for Buffer and GeoPoint data types (#83)"
sent on 12/21/15 9:11:38 AM.

This is the only notification you will receive while this person is away.

bajtos commented on LoopBackTests/LBPersistedModelSubclassingTests.m:
```diff
@@ -90,38 +95,63 @@ - (void)testRepositoryNotOverridden {
+        @"data" : @{ @"data1": @1, @"data2": @2 },
+        @"array": @[ @"one", @"two", @"three" ],
+        @"date" : @"1970-01-01T00:00:00.000Z",
+        @"buffer": @{ @"type": @"Buffer", @"data": @[ @12, @34, @56 ] },
```
This should be `@{ @"$type": @"base64", @"$data": @"DCI4" }`

bajtos replied:
^^^ Ignore that comment, it's based on a wrong assumption.

bajtos commented on LoopBackTests/LBPersistedModelSubclassingTests.m:
```diff
@@ -157,52 +187,67 @@ - (void)testAll {
+                XCTAssertEqual(widget.geopoint.coordinate.latitude, 12.3, @"Invalid latitude.");
+                XCTAssertEqual(widget.geopoint.coordinate.longitude, 45.6, @"Invalid longitude.");
+
+                // restore
```
Assuming that `XCTAssert*` throws on failure, this "restore" block will not be run when any of the assertion failed, thus triggering false failures in tests that depend on these values and running after this test.

You could move this code to some sort of "run after each test" hook. However, that would make troubleshooting of failed tests a bit difficult, because the database state would be cleared by the time you get to inspect it after tests finished. The solution is to run this "restore" code before each test - that way every test starts with a clean env and doesn't need clean up after itself.

hideya replied:
Good suggestion.  I didn't like the way it is currently handled neither.  I'm thinking to update the test not to depend on a model stored on the server, rather creates a temporary model and update it, so that we don't need to worry about target model's state before running the test.

raymondfeng replied:
I am out of the office until 01/04/2016.

Happy holidays!


Note: This is an automated response to your message  "Re:
[loopback-sdk-ios] Add support for Buffer and GeoPoint data types (#83)"
sent on 12/21/15 9:07:03 AM.

This is the only notification you will receive while this person is away.

bajtos replied:
> I'm thinking to update the test not to depend on a model stored on the server, rather creates a temporary model and update it, so that we don't need to worry about target model's state before running the test.

:+1: :clap:

raymondfeng replied:
I am out of the office until 01/04/2016.

Happy holidays!


Note: This is an automated response to your message  "Re:
[loopback-sdk-ios] Add support for Buffer and GeoPoint data types (#83)"
sent on 12/21/15 2:20:58 PM.

This is the only notification you will receive while this person is away.

bajtos commented on LoopBackTests/server/index.js:
```diff
@@ -68,7 +80,21 @@ Widget.destroyAll(function () {
+      'three'
+    ],
+    date: '1970-01-01T00:00:00.000Z',
+    buffer: [12, 34, 56],
```
Oh, so it looks like you did actually add a test where the `buffer` value is created/parsed by the server. This code is testing how LoopBack properties of type `Buffer` are handled. My comment above is referring to the way how strong-remoting handles method arguments of type `Buffer`.

bajtos commented:
Apparently we have two ways how to encode a buffer value:

1. Method/callback arguments are encoded as `{ '$type': 'base64', '$data': 'MjM=' }`
2. Model properties are encoded as `{ type: 'Buffer', data: [50, 51] }`

@hideya is there a reasonably simple way how to support both encodings? Base64-encoding would be triggered only for top-level properties in response request/body, array-based encoding would be triggered everywhere else.

I think we should eventually fix this inconsistency, see https://github.com/strongloop/loopback/issues/1907

hideya commented:
@bajtos thank you always for the careful review!

As to the two encodings of Buffer, I'll look into a way to support the both.  Thanks for the clarifications!

raymondfeng commented:
I am out of the office until 01/04/2016.

Happy holidays!


Note: This is an automated response to your message  "Re:
[loopback-sdk-ios] Add support for Buffer and GeoPoint data types (#83)"
sent on 12/21/15 8:56:26 AM.

This is the only notification you will receive while this person is away.

raymondfeng commented:
I am out of the office until 01/04/2016.

Happy holidays!


Note: This is an automated response to your message  "Re:
[loopback-sdk-ios] Add support for Buffer and GeoPoint data types (#83)"
sent on 12/21/15 8:19:33 AM.

This is the only notification you will receive while this person is away.

bajtos commented:
@hideya you are welcome. I'll reassign this PR back to you, so that it stays out of my list of PRs to review. Please assign it back to me when there is more to review.

hideya commented:
@bajtos  I found Data object is also encoded in two different ways:

1. Method/callback arguments: `{ '$type': 'date', '$data': 'Thu Jan 01 1970 09:00:00 GMT+0900 (JST)' }`
2. Model properties are encoded: `'1970-01-01T00:00:00.000Z'`

The format difference is caused because [the method/callback arg case uses `date.toString()`](https://github.com/strongloop/strong-remoting/blob/e0846fe0f9d5480eb3e2ec5105b5fe6b4b7fa332/lib/shared-method.js#L544-L547), whereas it looks `date.toISOString()` is used for properties (I couldn't figure out the exact line where the conversion is taken place).  Is my understanding correct?

hideya commented on LoopBack/LBModel.h:
```diff
@@ -8,6 +8,68 @@
+ * Blocks of this type are executed for a successful method invocation that doesn't return
+ * any value as a callback argument.
+ */
+typedef void (^LBModelVoidSuccessBlock)();
```
For convenience (especially for CodeGen), the success block types with supported data types are typedef'ed.

hideya commented on LoopBack/LBModel.m:
```diff
@@ -63,9 +54,6 @@ - (NSDictionary *)toDictionary {
             id obj = [self valueForKey:propertyName];
-            if ([obj isKindOfClass:[NSDate class]]) {
-                obj = [jsonDateFormatter stringFromDate:obj];
-            }
```
Parameter type conversion has been moved to `SLObject`

hideya commented on LoopBack/LBModel.m:
```diff
@@ -122,9 +110,21 @@ - (LBModel *)modelWithDictionary:(NSDictionary *)dictionary {
+                // if the property type is CLLocation, convert the json object to a location object
+                else if (strncmp(type, "T@\"CLLocation\",", 15) == 0) {
+                    obj = [SLObject locationFromEncodedProperty:obj];
+                }
```
The parameter conversions back to iOS data types are needed to be performed here in `LBModel` since it requires access to model property type attributes.

hideya commented on LoopBack/LBPersistedModel.h:
```diff
@@ -7,12 +7,13 @@
-typedef void (^LBPersistedModelArraySuccessBlock)(NSArray *array);
-typedef void (^LBPersistedModelVoidSuccessBlock)();
-typedef void (^LBPersistedModelBoolSuccessBlock)(BOOL boolean);
-typedef void (^LBPersistedModelNumberSuccessBlock)(NSInteger number);
```
Since those success block types rather belong to `LBModel`, they have been moved there.

bajtos replied:
Hmm, isn't this a breaking change? I mean if there is any existing user code relying on `LBPersistedModelVoidSuccessBlock` (for example), then it will break after upgrading to a new version of iOS SDK.

Can we keep these typedefs around, perhaps with a deprecation notice (if there is a mechanism for that).

```io
typedef LBModelVoidSuccessBlock LBPersistedModelVoidSuccessBlock;
```

hideya replied:
I didn't care about backward compatibility for those since those typedefs are added by me on Oct. 29 as a part of CodeGen support.  It seems highly unlikely that someone already depends on the typedefs.  How about make the above change and consider backward compatibility when we get any Issue on it?

hideya replied:
Following @bajtos 's suggestion, I've left the typedefs with deprecation warning since I'm informed that Ver. v1.3.2  has been recently which includes all the typedefs (which I didn't realise until yesterday).

hideya commented on LoopBackTests/LBPersistedModelSubclassingTests.m:
```diff
@@ -17,13 +17,7 @@
-@property (nonatomic) NSNumber *flag;
-@property BOOL flag2;
-@property (nonatomic) NSDate *date;
 
```
Data type tests have been moved into an independent test file `LBPersistedModelDataTypesTests.m`

hideya commented on LoopBackTests/LBPersistedModelTests.m:
```diff
@@ -52,17 +52,20 @@ - (void)tearDown {
 }
 
 - (void)testCreate {
-    LBPersistedModel __block *model = (LBPersistedModel*)[self.repository modelWithDictionary:@{ @"name": @"Foobar", @"bars": @1 }];
```
Fixed indentations etc. to follow other part of the code.

hideya commented on LoopBackTests/bin/run-unit-tests:
```diff
@@ -26,12 +26,19 @@ node LoopBackTests/server/index.js &
+# CI's xcode only has old simulators
+if [[ -n "${JENKINS_HOME}" ]]; then
+  DESTINATION_SIMULATOR=${DESTINATION_SIMULATOR_CI}
+fi
```
This is needed for running the tests on Xcode 7 (the latest) and 5 (on CI).

hideya commented:
Updated to deal with method/callback argument value encoding/decoding in addition to properties.

It turned out that Buffer and Date values for method/callback arguments are encoded in different ways  than for properties.  To deal with it, I've made a large changes to this PR and added unit tests to check method/callback argument encoding/decoding.  I think this PR should work fine.

@bajtos I'm assigning this back to you.  It would be great if you could review the changes.

bajtos commented on LoopBackTests/LBPersistedModelDataTypesTests.m:
```diff
@@ -0,0 +1,232 @@
+#import "SLRemotingTestsUtils.h"
+
+
+@interface Widget2 : LBPersistedModel
```
Please use a more descriptive name. What is the difference between `Widget` and `Widget2`?

hideya replied:
The reason why I needed to use the name `Widget2` is that the class name `Widget` has been used in a different unit test `LBPersistedModelSubclassingTests.m` and I needed to avoid conflict in names (note that `LBPersistedModelDataTypesTests.m` is separated from `LBPersistedModelSubclassingTests.m` for readability).  The difference between them is that `Widget2` defines more properties than `Widget` for testing property handling.

So, how about using `WidgetEx` instead of `Widget2`? (where `Ex` stands for extra properties)

hideya replied:
In order to avoid any confusion, I decided to merge `LBPersistedModelDataTypesTests.m` into `LBPersistedModelSubclassingTests.m` since the both files are testing against the same `Widget` model.

bajtos commented:
@hideya I have reviewed the patch, it looks pretty good now. I did not review the details thoroughly though, I'll trust the tests. There are few things to improve, see my comments above.

IIUC, the strong-remoting part implements correct serialization of date/buffer/geopoint values when they are passed as method arguments. However, the response ("returns" argument) is always passed as a JSON object and it's up to the API caller to extract individual arguments from the response, together with applying any type conversions as necessary. Is that correct? Makes me wonder how much work would be needed to fix this part too, so that the user provides typed input arguments and gets typed output arguments back. Thoughts?

hideya commented:
@bajtos thank you for the nice review!  I've added comments to answer your questions.

> the response ("returns" argument) is always passed as a JSON object and it's up to the API caller to extract individual arguments from the response, together with applying any type conversions as necessary. Is that correct?

Yes, it is.

> Makes me wonder how much work would be needed to fix this part too, so that the user provides typed input arguments and gets typed output arguments back. Thoughts?

Very good point!  I was also wondering if we could simplify that part. Actually, the same discussion can be applied to all the other data types including Model and Array.  Maybe we can simply follow Android's callback argument handling?  In any case, it looks to me better be done in a separate PR.

hideya commented:
@bajtos I've updated this by following our conversation.  Would you please review the changes once again?  Thank you always for your help!

bajtos commented:
@hideya LGTM, thank you!
